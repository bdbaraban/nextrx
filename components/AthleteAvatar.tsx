import React from 'react';
import { Box, Image, Text } from 'grommet';
import { Athlete } from '../db/types';

interface AthleteAvatarProps {
  athlete: Athlete;
}

const AthleteAvatar = ({ athlete }: AthleteAvatarProps): React.ReactElement => {
  return (
    <Box direction="row" align="center" gap="small">
      <Box alignSelf="center" height="xsmall" width="xsmall">
        <Image
          a11yTitle="Athlete profile picture"
          src={athlete.profile_image_url}
        />
      </Box>
      <Box justify="between">
        <Text weight="bold" size="large" truncate={true}>
          {`${athlete.first_name} ${athlete.last_name}`}
        </Text>
        <Text size="medium" truncate={true}>
          {athlete.affiliate.name}
        </Text>
      </Box>
    </Box>
  );
};

export default AthleteAvatar;
