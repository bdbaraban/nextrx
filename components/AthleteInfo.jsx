import React from 'react';
import { Box, Heading, Image } from 'grommet';

const AthleteInfo = props => {
  return (
    <React.Fragment>
      <Box
        alignSelf="center"
        height="xsmall"
        width="xsmall"
        margin={{ right: 'medium' }}
      >
        <Image src="https://scontent-sjc3-1.xx.fbcdn.net/v/t1.0-9/46857583_1939285332773307_9124487703423877120_n.jpg?_nc_cat=101&_nc_oc=AQkPTt0sxNWzpg_Pl0EsflTCE-GRCMiyDTkq_ZoAYkSeHKaCA38LJ3Fc-ajnotWOeqg&_nc_ht=scontent-sjc3-1.xx&oh=1373649fb4681b631eb4bb9fc39d8164&oe=5DC53061" />
      </Box>
      <Box direction="column">
        <Heading level="2" textAlign="center" margin="xsmall">
          Brennan D Baraban
        </Heading>
        <Box
          flex
          direction={props.size === 'small' ? 'column' : 'row'}
          justify="center"
          align="center"
        >
          <Heading level="4" textAlign="center" margin="xsmall">
            Affiliate:
          </Heading>
          <Heading level="4" textAlign="center" margin="xsmall">
            CrossFit Novato
          </Heading>
        </Box>
      </Box>
    </React.Fragment>
  );
};

export default AthleteInfo;
