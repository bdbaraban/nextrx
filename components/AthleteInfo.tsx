import React from 'react';
import { Box, Text } from 'grommet';
import { Athlete, Workout } from '../db/types';
import { Emoji } from '.';

// Months codex
const MONTHS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
];

/**
 * Calculate the previous Monday of a given date.
 * @param date - Date to calculate previous Monday of.
 */
const getMonday = (date: Date): Date => {
  const day = date.getDay();
  const diff = date.getDate() - day + (day === 0 ? -6 : 1);
  return new Date(date.setDate(diff));
};

/**
 * Count number of workouts posted within current week.
 * @param workouts - Array of total athlete workouts.
 */
const countActiveDays = (workouts: Workout[]): number => {
  let count = 0;
  const monday = getMonday(new Date());

  for (
    let i = workouts.length - 1, current = new Date(workouts[i].date);
    current > monday || current.getDate() === monday.getDate();
    current = new Date(workouts[--i].date)
  ) {
    count++;
  }

  return count;
};

interface AthleteInfoProps {
  athlete: Athlete;
}

/**
 * Athlete profile information
 */
const AthleteInfo = ({ athlete }: AthleteInfoProps): React.ReactElement => {
  const createdAt = new Date(athlete.created_at);
  const activeDays = countActiveDays(athlete.workouts);

  return (
    <Box direction="row" justify="between">
      <Box direction="column" justify="center" gap="xsmall">
        <Text size="medium">
          <Emoji label="tear-off-calendar" symbol="📆" /> Member since:{' '}
        </Text>
        <Text size="medium">
          <Emoji label="person-running" symbol="🏃" /> Active days this week:{' '}
        </Text>
        <Text size="medium">
          <Emoji label="flex-biceps" symbol="💪" /> Total logged WOD&apos;s:{' '}
        </Text>
      </Box>
      <Box direction="column" align="end" justify="center" gap="xsmall">
        <Text size="medium">
          {`${MONTHS[createdAt.getMonth() - 1]} ${createdAt.getFullYear()}`}
        </Text>
        <Text size="medium">{activeDays}</Text>
        <Text size="medium">{athlete.workouts.length}</Text>
      </Box>
    </Box>
  );
};

export default AthleteInfo;
