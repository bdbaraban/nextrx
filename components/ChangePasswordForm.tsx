import React from 'react';
import { Action } from 'redux';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { Box, Button, Form, FormField } from 'grommet';
import { updateAthlete } from '../store/athlete/actions';
import { AthleteState } from '../store/athlete/types';
import { AppState } from '../store';
import { Athlete } from '../db/types';
import { Popper } from '.';

/**
 * Mapped Redux state
 */
interface SettingsFormStateProps {
  athlete: Athlete;
  error: string;
}

/**
 * Mapped Redux actions
 */
interface SettingsFormDispatchProps {
  handleSubmit: (attributes: Partial<Athlete>) => void;
}

/**
 * SettingsForm component prop types
 */
type SettingsFormProps = SettingsFormStateProps & SettingsFormDispatchProps;

/**
 * Popper state type
 */
interface PopperState {
  open: boolean;
  message: string;
}

/**
 * Settings page form for changing account password.
 */
const ChangePasswordForm = ({
  athlete,
  error,
  handleSubmit
}: SettingsFormProps): React.ReactElement => {
  const [popper, setPopper] = React.useState<PopperState>({
    open: false,
    message: ''
  });

  const flashPopper = (message: string): void => {
    setPopper({
      open: true,
      message
    });
    setTimeout((): void => {
      setPopper({
        open: false,
        message: ''
      });
    }, 3500);
  };

  const onSubmit = (event: any): void => {
    handleSubmit({ _id: athlete._id, ...event.value });
    flashPopper(error === '' ? 'Changes saved!' : error);
  };

  return (
    <Form onSubmit={onSubmit}>
      <FormField label="Current Password" name="currentPassword" />
      <FormField label="New Password" name="newPassword" />
      <FormField label="Confirm New Password" name="confirmPassword" />
      <Box pad="small">
        <Button label="Save Changes" type="submit" />
        {popper.open && <Popper message={popper.message} />}
      </Box>
    </Form>
  );
};

const mapStateToProps = (state: AppState): SettingsFormStateProps => {
  return {
    athlete: state.athlete.profile,
    error: state.athlete.error
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<AthleteState, null, Action<string>>
): SettingsFormDispatchProps => {
  return {
    handleSubmit: (attributes: Partial<Athlete>): void => {
      dispatch(updateAthlete(attributes));
    }
  };
};

export default connect<
  SettingsFormStateProps,
  SettingsFormDispatchProps,
  {},
  AppState
>(
  mapStateToProps,
  mapDispatchToProps
)(ChangePasswordForm);
