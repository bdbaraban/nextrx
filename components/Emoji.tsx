import React from 'react';

interface EmojiProps {
  label: string;
  symbol: string;
}

const Emoji = ({ label, symbol }: EmojiProps): React.ReactElement => (
  <span
    className="emoji"
    role="img"
    aria-label={label ? label : ''}
    aria-hidden={label ? 'false' : 'true'}
  >
    {symbol}
  </span>
);

export default React.memo(Emoji);
