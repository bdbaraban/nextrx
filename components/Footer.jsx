import React from 'react';
import { Box, Text } from 'grommet';

const Footer = () => {
  return (
    <Box
      tag="footer"
      direction="row"
      align="center"
      justify="center"
      background="light-2"
      pad={{ left: 'medium', right: 'small', vertical: 'small' }}
      margin={{ top: 'medium' }}
    >
      <Text level="3" margin="none">
        RXAthlete Inc.
      </Text>
    </Box>
  );
};

export default React.memo(Footer);
