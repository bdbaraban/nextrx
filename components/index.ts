// Components barrel file
export { default as AthleteAvatar } from './AthleteAvatar';
export { default as ActivityChart } from './ActivityChart';
export { default as SidebarMenu } from './SidebarMenu';
export { default as ThemeToggleButton } from './ThemeToggleButton';
export { default as Title } from './Title';
