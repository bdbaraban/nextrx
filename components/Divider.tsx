import React from 'react';
import { connect } from 'react-redux';
import { Box, Heading } from 'grommet';
import { ThemeState } from '../store/theme/types';
import { AppState } from '../store';
import Emoji from './Emoji';

interface DividerProps {
  theme: ThemeState;
  symbol: string;
  text: string;
  label: string;
}

/**
 * Emoji-styled section divider.
 */
const Divider = ({
  theme,
  symbol,
  text,
  label
}: DividerProps): React.ReactElement => {
  return (
    <Box fill="horizontal" align="center" justify="center">
      <Heading
        textAlign="center"
        level="4"
        style={
          theme.type === 'dark'
            ? {
                borderBottom: '2px dotted #666666',
                lineHeight: 0,
                minWidth: '85%'
              }
            : {
                borderBottom: '2px dotted #000',
                lineHeight: 0,
                minWidth: '85%'
              }
        }
      >
        <span
          style={
            theme.type === 'dark'
              ? {
                  background: '#181a1b'
                }
              : {
                  background: '#fff'
                }
          }
        >
          <Emoji label={label} symbol={symbol} />
          {` ${text} `}
          <Emoji label={label} symbol={symbol} />
        </span>
      </Heading>
    </Box>
  );
};

const mapStateToProps = (state: AppState): { theme: ThemeState } => {
  return { theme: state.theme };
};

export default connect(mapStateToProps)(Divider);
