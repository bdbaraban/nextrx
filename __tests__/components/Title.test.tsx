import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Link from 'next/link';
import { Heading } from 'grommet';
import { Logo, Title } from '../../components';

describe('Title', (): void => {
  it('Links to /athlete route', (): void => {
    const title = shallow(<Title />);
    expect(title.find(Link).prop('href')).toEqual('/athlete');
  });

  it('Renders Logo', (): void => {
    const title = shallow(<Title />);
    expect(title.find(Logo)).toHaveLength(1);
  });

  it('Renders Heading with "NextRX" title', (): void => {
    const title = shallow(<Title />);
    expect(
      title
        .find(Heading)
        .render()
        .text()
    ).toEqual('NextRX');
  });

  it('Matches snapshot', (): void => {
    const component = renderer.create(<Title />);
    const title = component.toJSON();
    expect(title).toMatchSnapshot();
  });
});
