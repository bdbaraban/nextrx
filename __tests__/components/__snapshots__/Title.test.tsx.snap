// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title Matches snapshot 1`] = `
<a
  className="jsx-3033568672"
  href="/athlete"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="StyledBox-sc-13pk1d4-0 jCbcwt"
  >
    <svg
      fill="#fff"
      height={42}
      viewBox="0 0 160 205"
      width={42}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M54.715 146.293H42.7V167h-6.785v-51.188H52.85c5.766 0 10.195 1.313 13.3 3.938 3.117 2.625 4.676 6.445 4.676 11.46 0 3.187-.867 5.965-2.602 8.332-1.7 2.367-4.102 4.137-7.172 5.3l12.023 21.726V167h-7.242l-11.1-20.707zm-12.023-5.52h10.37c3.352 0 6.012-.867 7.98-2.6 1.992-1.734 2.988-4.055 2.988-6.96 0-3.164-.95-5.6-2.848-7.277-1.875-1.688-4.594-2.543-8.156-2.567H42.7v19.406zM156.312 205h-17.296L21.313 29.78V205H3.875V.25h17.438L139.157 175.6V.25h17.156V205zM107.64 51.43l12.27-19.617h7.945l-16.17 25.383L128.242 83h-8.015L107.64 63.03 94.984 83h-7.98l16.594-25.805L87.4 31.813h7.9l12.34 19.617z"
      />
    </svg>
    <h1
      className="StyledHeading-sc-1rdh4aw-0 kBxQZt"
      size="small"
    >
      NextRX
    </h1>
  </div>
</a>
`;
